rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+
+## For the cellxgene Data Portal
+
+**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)
+
+**Approvers:** [Arathi Mani](mailto:arathi.mani@chanzuckerberg.com), [Timmy Huang](mailto:thuang@chanzuckerberg.com),
+[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com), [Eduardo Lopez](mailto:elopez@chanzuckerberg.com), [Brian Raymor](mailto:braymor@chanzuckerberg.com)
+[Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)
+
+## tl;dr
+
+A user can upload files, up to 30GB in size, to Data Portal from a DropBox link.
+
+## Glossary
+
+- Shared Link - A link from DropBox that gives the bearer of that link access to private data.
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 100 MB in size.
+- Large File - a file > 100 MB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own files. These files consist of data sets.
+These data sets can vary in size from a few MB to several GB. Users can provide a shared link from their DropBox that
+is used to upload directly to S3.
+
+This specification does not cover the validation of the data or the process for a user to complete a submission.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their Dropbox.
+1. A user can initiate an upload from the DP Browser App.
+1. A user can abort an upload from the DP Browser App.
+1. A user can check the status of their upload from the DP Browser App.
+1. A user can restart a failed upload from the DP Browser App.
+1. The user must be authenticated before uploading.
+1. A user who has not yet agreed to the Portal policies must not be able to upload their file.
+
+### Nonfunctional requirements
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
+1. Shared links referring to more than one file will be rejected.
+1. The status of an upload must be tracked and queryable.
+1. Only .h5ad files are accepted.
+
+## Detailed Design | Architecture | Implementation
+
+![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
+Figure 1: The Upload Design
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their DropBox from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 3.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 3:** Architecture of a user uploading from their cloud to our cloud.
+
+#### 1. Get Share Link
+
+The user must generate a sharable link from their DropBox. Sharable links should only point to a single file, not a
+directory of files.
+
+#### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
```

Comment:
Is this a `POST` or a `PUT`?

Reply:
POST since we are not adding something to this resource path. The resource we are uploading will appear in /dataset/{dataset_id}.